ToolBox/CoinApi/CoinApiDataQueueHandler.cs

```diff
@@ -0,0 +1,360 @@
+        private readonly WebSocketWrapper _webSocket = new WebSocketWrapper();
+        private readonly object _locker = new object();
+        private readonly List<Tick> _ticks = new List<Tick>();
+        private readonly DefaultConnectionHandler _connectionHandler = new DefaultConnectionHandler();
+        private readonly CoinApiSymbolMapper _symbolMapper = new CoinApiSymbolMapper();
+
+        private readonly TimeSpan _subscribeDelay = TimeSpan.FromMilliseconds(250);
+        private readonly object _lockerSubscriptions = new object();
+        private HashSet<Symbol> _subscribedSymbols = new HashSet<Symbol>();
+        private DateTime _lastSubscribeRequestUtcTime = DateTime.MinValue;
+        private bool _subscriptionsPending;
+
+        private List<string> _subscribedExchanges = new List<string>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CoinApiDataQueueHandler"/> class
+        /// </summary>
+        public CoinApiDataQueueHandler()
+        {
+            _connectionHandler.ConnectionLost += OnConnectionLost;
+            _connectionHandler.ConnectionRestored += OnConnectionRestored;
+            _connectionHandler.ReconnectRequested += OnReconnectRequested;
+
+            _connectionHandler.Initialize();
+
+            _webSocket.Initialize(WebSocketUrl);
+
+            _webSocket.Message += OnMessage;
+
+            _webSocket.Connect();
+        }
+
+        /// <summary>
+        /// Get the next ticks from the live trading data queue
+        /// </summary>
+        /// <returns>IEnumerable list of ticks since the last update.</returns>
+        public IEnumerable<BaseData> GetNextTicks()
+        {
+            lock (_locker)
+            {
+                var copy = _ticks.ToArray();
+                _ticks.Clear();
+                return copy;
+            }
+        }
+
+        /// <summary>
+        /// Adds the specified symbols to the subscription
+        /// </summary>
+        /// <param name="job">Job we're subscribing for:</param>
+        /// <param name="symbols">The symbols to be added keyed by SecurityType</param>
+        public void Subscribe(LiveNodePacket job, IEnumerable<Symbol> symbols)
+        {
+            lock (_lockerSubscriptions)
+            {
+                var symbolsToSubscribe = (from symbol in symbols
+                                          where !_subscribedSymbols.Contains(symbol) && CanSubscribe(symbol)
+                                          select symbol).ToList();
+                if (symbolsToSubscribe.Count == 0)
+                    return;
+
+                Log.Trace($"CoinApiDataQueueHandler.Subscribe(): {string.Join(",", symbolsToSubscribe.Select(x => x.Value))}");
+
+                // CoinAPI requires at least 5 seconds between subscription requests so we need to batch them
+                _subscribedSymbols = symbolsToSubscribe.Union(_subscribedSymbols).ToHashSet();
```

Comment:
nit - no need to change - could use `Concat` here since `ToHashSet` will perform the set union as well.